File: backend/schemas/cad_pipeline.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

CadPipelineStepName = Literal[
    "convert_dwg",
    "prepare_dxf_sheet",
    "parse_dxf",
    "generate_candidates",
    "fuse_fields",
    "generate_cad_preview",
]

CadPipelineStatus = Literal["success", "completed_with_errors", "failed", "skipped"]


class CadPipelineRequest(BaseModel):
    steps: list[CadPipelineStepName] = Field(
        default_factory=lambda: [
            "convert_dwg",
            "prepare_dxf_sheet",
            "parse_dxf",
            "generate_candidates",
            "fuse_fields",
        ]
    )
    skip_completed: bool = True
    continue_on_error: bool = True


class CadPipelineError(BaseModel):
    file_id: int | None = None
    sheet_id: int | None = None
    file_name: str | None = None
    step: CadPipelineStepName
    error_code: str
    message: str


class CadPipelineItem(BaseModel):
    file_id: int | None = None
    sheet_id: int | None = None
    file_name: str | None = None
    status: str
    error_code: str | None = None
    message: str | None = None


class CadPipelineStepResult(BaseModel):
    step: CadPipelineStepName
    status: CadPipelineStatus
    duration_seconds: float = 0
    success_count: int = 0
    failed_count: int = 0
    skipped_count: int = 0
    warning_count: int = 0
    items: list[CadPipelineItem] = Field(default_factory=list)
    errors: list[CadPipelineError] = Field(default_factory=list)


class CadPipelineSummary(BaseModel):
    duration_seconds: float = 0
    start_time: datetime | None = None
    finish_time: datetime | None = None
    total_files: int
    pdf_files: int = 0
    dwg_files: int
    dxf_files: int
    converted_success: int = 0
    converted_failed: int = 0
    sheet_prepared_success: int = 0
    sheet_prepared_failed: int = 0
    parse_success: int = 0
    parse_failed: int = 0
    candidate_success: int = 0
    candidate_failed: int = 0
    fusion_success: int = 0
    fusion_failed: int = 0
    cad_preview_success: int = 0
    cad_preview_failed: int = 0
    cad_preview_skipped: int = 0
    cad_preview_warning_count: int = 0
    skipped_count: int = 0
    error_count: int = 0
    warning_count: int = 0


class CadPipelineResponse(BaseModel):
    batch_id: int
    status: CadPipelineStatus
    summary: CadPipelineSummary
    steps: list[CadPipelineStepResult]
    errors: list[CadPipelineError] = Field(default_factory=list)

File: frontend/src/api/cadPipeline.ts
import { apiGet, apiPost } from "./client";

export type CadPipelineStep =
  | "convert_dwg"
  | "prepare_dxf_sheet"
  | "parse_dxf"
  | "generate_candidates"
  | "fuse_fields"
  | "generate_cad_preview";

export type CadPipelineRequest = {
  steps: CadPipelineStep[];
  skip_completed: boolean;
  continue_on_error: boolean;
};

export type CadPipelineError = {
  file_id: number | null;
  sheet_id: number | null;
  file_name: string | null;
  step: CadPipelineStep;
  error_code: string;
  message: string;
};

export type CadPipelineStepResult = {
  step: CadPipelineStep;
  status: string;
  duration_seconds: number;
  success_count: number;
  failed_count: number;
  skipped_count: number;
  warning_count: number;
  items: Array<{
    file_id: number | null;
    sheet_id: number | null;
    file_name: string | null;
    status: string;
    error_code: string | null;
    message: string | null;
  }>;
  errors: CadPipelineError[];
};

export type CadPipelineResponse = {
  batch_id: number;
  status: string;
  summary: {
    duration_seconds: number;
    start_time: string | null;
    finish_time: string | null;
    total_files: number;
    pdf_files: number;
    dwg_files: number;
    dxf_files: number;
    converted_success: number;
    converted_failed: number;
    sheet_prepared_success: number;
    sheet_prepared_failed: number;
    parse_success: number;
    parse_failed: number;
    candidate_success: number;
    candidate_failed: number;
    fusion_success: number;
    fusion_failed: number;
    cad_preview_success: number;
    cad_preview_failed: number;
    cad_preview_skipped: number;
    cad_preview_warning_count: number;
    skipped_count: number;
    error_count: number;
    warning_count: number;
  };
  steps: CadPipelineStepResult[];
  errors: CadPipelineError[];
};

export type BackgroundJobStatus = {
  id: number;
  job_type: string;
  scope_type: "batch" | "project" | "sheet";
  scope_id: number;
  status: "running" | "completed" | "failed";
  total: number;
  processed: number;
  current_step: string | null;
  message: string | null;
  started_at: string;
  finished_at: string | null;
  result_summary: CadPipelineResponse | null;
};

export function startCadPipeline(
  batchId: number,
  payload: CadPipelineRequest
): Promise<BackgroundJobStatus> {
  return apiPost<BackgroundJobStatus>(`/api/imports/${batchId}/cad-pipeline`, payload);
}

export function getCadPipelineJob(
  batchId: number
): Promise<BackgroundJobStatus | null> {
  return apiGet<BackgroundJobStatus | null>(`/api/imports/${batchId}/cad-pipeline/job`);
}
